opr/grading.go

```diff
@@ -207,27 +207,30 @@ func GetEntryBlocks(config *config.Config) {
 	// the next block we are going to process.
 	// Ignore opr blocks that don't get 10 winners.
 	for i := len(oprblocks) - 1; i >= 0; i-- { // Okay, go through these backwards
-		prevOPRBlock := GetPreviousOPRs(int32(oprblocks[i].Dbht)) // Get the previous OPRBlock
-		var validOPRs []*OraclePriceRecord                        // Collect the valid OPRPriceRecords here
-		for _, opr := range oprblocks[i].OPRs {                   // Go through this block
-			for j, eh := range opr.WinPreviousOPR { // Make sure the winning records are valid
-				if (prevOPRBlock == nil && eh != "") ||
-					(prevOPRBlock != nil && eh != prevOPRBlock[0].WinPreviousOPR[j]) {
-					continue
-				}
-				opr.Difficulty = opr.ComputeDifficulty(opr.Nonce)
-
-				// TODO: Enforce this?
-				f := binary.BigEndian.Uint64(opr.SelfReportedDifficulty)
-				if f != opr.Difficulty {
-					//This is a falsely reported difficulty. There is nothing we can
-					//really do. Maybe we should log.warn how many per block are 'malicious'?
-					log.Errorf("Diff mistmatch. Exp %d, found %d", opr.Difficulty, f)
-				}
+		var validOPRs []*OraclePriceRecord // Collect the valid OPRPriceRecords here
+
+		var previousWinners []*OraclePriceRecord
+		prevblock := GetPreviousOPRBlock(int32(oprblocks[i].Dbht))
+		if prevblock != nil {
+			previousWinners = prevblock.GradedOPRs
+		}
+
+		for _, opr := range oprblocks[i].OPRs { // Go through this block
+			if !VerifyWinners(opr, previousWinners) {
+				continue
+			}
+			opr.Difficulty = opr.ComputeDifficulty(opr.Nonce)
 
+			f := binary.BigEndian.Uint64(opr.SelfReportedDifficulty)
+			if f != opr.Difficulty {
+				// TODO Maybe we should log.warn how many per block are 'malicious'?
+				log.Errorf("Diff mistmatch. Exp %d, found %d", opr.Difficulty, f)
```

Comment:
Can we log more context? The ehash and height?

I know I wrote this, but if we are enforcing it, more context would help.
My rational behind the error, was that in the testnet we shouldn't have anyone malicious (unless we mess with ourselves). So if we report this error, there is a bug in our miner/grader that needs addressing. Once we go live, I'm ok with silencing this.

Reply:
i thought about it but the rest of the "self reported difficulty" functionality is not actually in yet. i just wanted to limit this PR to fixing the bugs mentioned, and then someone can build on top of it and implement the self-reported feature properly

Reply:
That's fair

Reply:
Ok, that is fine